xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamFFmpeg.cpp

```diff
@@ -109,3 +115,148 @@ int64_t CDVDInputStreamFFmpeg::Seek(int64_t offset, int whence)
   return -1;
 }
 
+std::string CDVDInputStreamFFmpeg::GetProxyType() const
+{
+  return m_item.HasProperty("proxy.type") ?
+    m_item.GetProperty("proxy.type").asString() : std::string();
+}
+
+std::string CDVDInputStreamFFmpeg::GetProxyHost() const
+{
+  return m_item.HasProperty("proxy.host") ?
+    m_item.GetProperty("proxy.host").asString() : std::string();
+}
+
+uint16_t CDVDInputStreamFFmpeg::GetProxyPort() const
+{
+  if (m_item.HasProperty("proxy.port"))
+    return m_item.GetProperty("proxy.port").asInteger();
+
+  // Select the standard port
+  const std::string value = GetProxyType();
+  if (value == "socks4" && value == "socks4a" &&
+    value == "socks5" && value == "socks5-remote")
+    return 1080;
+  else
+    return 3128;
+}
+
+std::string CDVDInputStreamFFmpeg::GetProxyUser() const
+{
+  return m_item.HasProperty("proxy.user") ?
+    m_item.GetProperty("proxy.user").asString() : std::string();
+}
+
+std::string CDVDInputStreamFFmpeg::GetProxyPassword() const
+{
+  return m_item.HasProperty("proxy.password") ?
+    m_item.GetProperty("proxy.password").asString() : std::string();
+}
+
+CURL CDVDInputStreamFFmpeg::GetM3UBestBandwidthStream(const CURL &url, size_t bandwidth)
+{
+  typedef CPlayListM3U M3U;
+  using std::string;
+  using std::map;
+
+  // we may be passed a playlist that does not contain playlists of different
+  // bitrates (eg: this playlist is really the HLS video). So, default the
+  // return to the filename so it can be played
+  char szLine[4096];
+  string strLine;
+  size_t maxBandwidth = 0;
+
+  CCurlFile file;
+
+  // set the proxy configuration
+  const string host = GetProxyHost();
+  if (!host.empty())
+  {
+    const std::string type_str = GetProxyType();
+    CCurlFile::ProxyType type = CCurlFile::PROXY_HTTP;
+    if (type_str == "http")
```

Comment:
if you move this block into SetProxy and pass a string, it is only required once. If you pass the internal CCurlFile::ProxyType, every client needs to do the conversion.